Use EFCore 3.1

lindellhugo opened the pull request:
Part of #323

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/App.config:
```diff
@@ -0,0 +1,13 @@
+      <provider invariantName="System.Data.SqlClient" type="System.Data.Entity.SqlServer.SqlProviderServices, EntityFramework.SqlServer" />
+    </providers>
+  </entityFramework>
+</configuration>
```
you can probaly drop this file, here are EF 6 references

Daniel-Svensson commented on src/Test/Desktop/EFCodePOCOModels/Northwind.edmx:
```diff
@@ -0,0 +1,158 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<edmx:Edmx Version="3.0" xmlns:edmx="http://schemas.microsoft.com/ado/2009/11/edmx">
```
I dont think we need to add any edmx file, the singel code first model in EfCoreModels project should be enough

Daniel-Svensson commented:
Some general review suggestions: @lindellhugo  @erikoijwall 

Read the issue for notes from previous discussions

- [ ] Manually diff EFCore project files against corresponding EF6 code
  (files should be very similar except where "model/metadata" inspection 
  where EFCore use IModel and EF6 use edmx/metadata workspace.
  - [ ] Handle TODO's
  - [ ] Remove unused or commented out resources and code which is not needed 
    (i've tried to remove "duplicates" and link to existing files instead)
- [ ] Manuall diff EFCore test related code agains corresponing ef6 
  (EFCoreModel project, baselines, Northwind/Catalog services ..) 
- Naming
    Try to keep names short, simple and consistent
      Eg. possible names for EFCore's DbDomainService can be called DbDomainService/EFCoreDomainService/EFDomainService
  - [ ] Make sure namespaces are in sync with project/nuget namespaces
    (private code code shared with other project can be kept in Server namespace)
  - [ ] Go thought names of all public types, methods and parameters
  - [ ] Make sure project files/paths match their names
* Look at "backporting" improvements in EFCore's DbDomainService to EF6 
version
   - [ ] Submit conflicts does not require "RefreshContext"
   - [ ] Submit conflicts can be made async
- Go though code generation tests (baselines) and make copies of all EF variants
  (e.g when EF6 has Northwind test for ObjectContext, POCO, DbContext and/or DbContext Code first) 
  a single EFCore test would be relevant.
  Look under `Baselines` folder of `OpenRiaServices.Common.DomainServices.Test` project. Make sure there are copies of relevant EF tests.
    Generated code will diff on namespace and names for `AssociationAttribute`s but should be similiar (if not identical) otherwise
    - [ ] Northwind
    - [ ] Catalog ("EF" folder ) 
    - [ ] EF_Inheritance
    - [ ] have I missed any ?
- [ ] Changes to VisualStudio related files can be included in a later PR

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/EFCoreTypeDescriptionContext.cs:
```diff
@@ -0,0 +1,153 @@
+
+        public IEntityType GetEntityType(Type type) => Model.FindEntityType(type);
+
+        ///// <summary>
```
remvoe both commented out methods ..

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/MetadataResource.resx:
```diff
@@ -0,0 +1,123 @@
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="LinqToEntitiesProvider_UnableToRetrieveMetadata" xml:space="preserve">
```
can this resource file be removed ?

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/TypeDescriptionContextBase.cs:
```diff
@@ -0,0 +1,51 @@
+    /// <summary>
+    /// Base class for LTS and EF type description contexts
+    /// </summary>
+    internal abstract class TypeDescriptionContextBase
```
efcore has only whan typedescriptor, move code to where it is used and remove file ?

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Test/DbContextModel/Northwind/NorthwindPartialClasses.cs:
```diff
@@ -0,0 +1,128 @@
+namespace EFCoreModels.Northwind
+{
+     // TODO: Remove MetaDataType attributes and move attributes directly to classe so project can be netstandard2
+    [MetadataType(typeof(RegionMetadata))]
```
determine if todo should be done before or after pr is merged

Daniel-Svensson commented on src/OpenRiaServices.Server/Test/OpenRiaServices.Server.Test.csproj:
```diff
@@ -14,6 +24,7 @@
     </Reference>
     <Reference Include="System.ComponentModel.DataAnnotations" />
     <Reference Include="System.configuration" />
+    <Reference Include="System.Data.Entity" />
```
I belive this is EF4 so it should be removed

Daniel-Svensson commented on src/OpenRiaServices.Server/Test/OpenRiaServices.Server.Test.csproj:
```diff
@@ -4,8 +4,18 @@
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.20" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.20">
```
are both tools and design needed ?  or do we inherit all required packages ?

Daniel-Svensson commented on src/OpenRiaServices.Tools/Test/CodeGenDomainServices.cs:
```diff
@@ -166,6 +166,30 @@ public void TestClientCodegen_EFDbCtxDomainServices_FullTypes()
+            string[] sharedFiles = Array.Empty<string>();
+
+            // Full type names
+            TestHelper.ValidateCodeGen(new TestHelper.CodeGenValidationOptions(@"FullTypeNames\Scenarios", "CG_Scenarios_EFDbContext_FullTypes", "EFDbContextScenarios.g",
```
maube files should be named somewhing with efcore

Daniel-Svensson commented on src/OpenRiaServices.Tools/Test/CodeGenDomainServices.cs:
```diff
@@ -166,6 +166,30 @@ public void TestClientCodegen_EFDbCtxDomainServices_FullTypes()
+        }
+
+        [DeploymentItem(@"Baselines\FullTypeNames\Scenarios", "CG_Scenarios_EFDbContext_FullTypes")]
+        [DeploymentItem(@"ProjectPath.txt", "CG_Scenarios_EFDbContext_FullTypes")]
```
you should be able to reuse same folder name (second parameter) as above to not get to manu extra testfolders

Daniel-Svensson commented on src/Test/Desktop/EFCodePOCOModels/App.Config:
```diff
@@ -0,0 +1,23 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
I belive this whole folder is unused and can be removed

Daniel-Svensson reviewed (commented):
partial review done

lindellhugo commented on src/OpenRiaServices.Client/Test/Client.Test/Data/CrossDomainServiceQueryTests.cs:
```diff
@@ -584,4 +584,19 @@ protected override Northwind CreateNorthwind()
             return new Northwind(TestURIs.DbCtx_Northwind);
         }
     }
+    /** TODO: 
```
Fix this.

lindellhugo commented on src/OpenRiaServices.Client/Test/Client.Test/Data/UpdateTests.cs:
```diff
@@ -16,6 +16,7 @@ namespace OpenRiaServices.Client.Test
     using Resource = SSmDsClient::OpenRiaServices.Client.Resource;
 
     /// <summary>
+    /// TODO: Make it work with EF Core. Maybe copy whole file and work with DBContext for EFCore
```
Fix this or remove

lindellhugo commented on src/OpenRiaServices.EntityFramework/Framework/DbDomainService.cs:
```diff
@@ -243,70 +241,55 @@ private async Task<bool> InvokeSaveChangesAsync(bool retryOnConflict, Cancellati
-                        {
-                            membersInConflict.Add(pd.Name);
+                            // TÒDO: Determine if Excluded properties should be skipped
+                            PropertyDescriptorCollection propDescriptors = TypeDescriptor.GetProperties(operationInConflict.Entity.GetType());
```
Fixed

lindellhugo commented on src/OpenRiaServices.Hosting.Wcf.OData/Test/OpenRiaServices.Hosting.Wcf.OData.Test.csproj:
```diff
@@ -35,6 +35,7 @@
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\..\OpenRiaServices.EntityFramework\Framework\OpenRiaServices.EntityFramework.csproj" />
+    <ProjectReference Include="..\..\OpenRiaServices.Hosting.Wcf.Endpoint\Test\OpenRiaServices.Hosting.Wcf.Endpoint.Test.csproj" />
```
Could be romoved so removed it.

Daniel-Svensson commented on src/OpenRiaServices.Server/Test/DomainServiceDescriptionTest.cs:
```diff
@@ -378,6 +387,19 @@ public void DomainServiceDescription_EFPOCO()
+            var ctxt = new EFCoreModels.Northwind.EFCoreDbCtxNorthwindEntities();
+            IEntityType entityType = ctxt.Model.FindEntityType(typeof(EFCoreModels.Northwind.Product).FullName);
+            Assert.IsNotNull(entityType);
+        }
```
remove this test? it only tests efcore not the provider

lindellhugo reviewed (commented):
Own review

Daniel-Svensson commented on src/OpenRiaServices.EntityFramework/Framework/Resource.Designer.cs:
```diff
@@ -19,10 +19,10 @@ namespace OpenRiaServices.EntityFramework {
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
     [global::System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
-    internal class Resource {
+    public class Resource {
```
Resource should probably not be made public

erikoijwall commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/EFCoreTypeDescriptor.cs:
```diff
@@ -0,0 +1,251 @@
+                _timestampProperty = timestampMembers[0];
+            }
+
+            // if (edmType.BuiltInTypeKind == BuiltInTypeKind.EntityType)
```
should this be used? otherwise might want to remove { }

erikoijwall replied:
There is also a TODO on foreignKeyMembers to change it to a List or HashSet

Daniel-Svensson replied:
I am ok with keeping this commented out code here (and maybe some other cases) as a reminder that there might be missing logic here, untill I know what other values of "edmType.BuiltInTypeKind" this check protects against for EF6.

It is very possible that the check is not needed and once it is understood why, it would be good to remove the code

Daniel-Svensson replied:
ok, for the checks in this file it seems like it excludes the case where `edmType.BuiltInTypeKind == BuiltInTypeKind.ComplexType`.

Daniel-Svensson replied:
it seems [owned entities](https://docs.microsoft.com/en-us/ef/core/modeling/owned-entities)  corresponds to ComplexTypes, I am fine with having an Issue for supporting "owned entities" or at least a TODO.md /README.md in the efcore project about this not yet beeing supported and some hints on what to change / where to look when supporing it.

I did however make a commit with some preliminary code for this, but adding final support for it would mean adding some units tests similar to `DomainServiceDescriptionTest.ComplexType_EFProviderTest` and related tests

lindellhugo replied:
Documented under "Not supported / limited support"

erikoijwall commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/EFCoreTypeDescriptor.cs:
```diff
@@ -0,0 +1,251 @@
+            //TODO: remove this, EntityState is not part of entities 
+
+            // exclude EntityState members
+            if (pd.PropertyType == typeof(EntityState)) // TODO: Maybe also check pd.Component type
```
Should pd.ComponentType be checked?

erikoijwall replied:
Or should all of this be removed, the comment above suggests that entitystate is not part of enties

Daniel-Svensson replied:
I think the comment on row 239 can be saftly removed, or merged with 241 to clarify that it is `IChangeDetector.EntityState` we want to exclude.
 EF core does support an EntityState property when implementing the `IChangeDetector` interface and it should not be considered part of the model / datacontract.


```suggestion
            if (pd.PropertyType == typeof(EntityState))
```

lindellhugo replied:
Looks good!

Daniel-Svensson commented on src/OpenRiaServices.EntityFramework/Framework/OpenRiaServices.EntityFramework.csproj:
```diff
@@ -1,7 +1,7 @@
   <PropertyGroup>
     <TargetFrameworks>net472;net6.0</TargetFrameworks>
-    <DefineConstants>$(DefineConstants);SERVERFX;DBCONTEXT;</DefineConstants>
+    <DefineConstants>$(DefineConstants);SERVERFX;DBCONTEXT;net472;</DefineConstants>
```
I dont think we want this define, it would add "net472" even for NET6 target

Daniel-Svensson commented on src/OpenRiaServices.EntityFramework/Framework/OpenRiaServices.EntityFramework.csproj:
```diff
@@ -43,5 +43,9 @@
       <LastGenOutput>MetadataResource.Designer.cs</LastGenOutput>
       <CustomToolNamespace>System.Data.Mapping</CustomToolNamespace>
     </EmbeddedResource>
+    <EmbeddedResource Update="Resource.resx">
```
I think this change should be removed or it will make the resource public

lindellhugo replied:
Removed

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/EFCoreTypeDescriptionContext.cs:
```diff
@@ -0,0 +1,98 @@
+            {
+                if (_model == null)
+                {
+                    // TODO: Is there a smarter way ?? 
```
the TODO comment makes some sense, but I aggree that it does not to say TODO .

I did not find a better way (maybe allow a second type parameter for model class och similar?) , unless you use compiled models which is only availible in latest versions of ef core

lindellhugo replied:
Rewrote todo to better state what can be done

Daniel-Svensson commented on src/OpenRiaServices.Server/Test/OpenRiaServices.Server.Test.csproj:
```diff
@@ -4,8 +4,12 @@
     <TargetFramework>net472</TargetFramework>
   </PropertyGroup>
+    <PropertyGroup>
+        <GenerateRuntimeConfigurationFiles>True</GenerateRuntimeConfigurationFiles>
```
what does this ssetting do ?

lindellhugo replied:
It generates files that specific .net Core runtime settings and version. Removed since we do not need it for net472.
https://stackoverflow.com/questions/44174407/what-is-the-purpose-of-msbuilds-generateruntimeconfigurationfiles

Daniel-Svensson commented on src/OpenRiaServices.Tools/Test/TestHelper.cs:
```diff
@@ -95,6 +95,7 @@ public static void ValidateFilesEqual(string relativeTestDir, string relativeDep
 
                 // Generate CMD strings to diff and to copy
                 string tfDiffCommand = "tf diff \"" + referenceFileName + "\" \"" + generatedFileName + "\"\r\n";
```
feel free to remove any mentions of "tf" commands (including edit) since only git is relevant

Daniel-Svensson commented on src/VisualStudio/Tools/Test/DomainServiceWizard/BusinessLogicClassContextTests.cs:
```diff
@@ -116,6 +117,18 @@ public void BusinessLogicClass_Context_EFDbContextTest()
+        //    //Assert.IsTrue(dbContext.NeedToGenerateMetadataClasses);
+        //    throw new NotImplementedException();
+        //}
+
```
does it make sense to keep the code ? or is it better to just remove the modification?

lindellhugo replied:
I think it makes sense since the TODO explains was must be done.
So I will keep it.

Daniel-Svensson commented on src/VisualStudio/Tools/Test/TestUtilities/TestHelper.cs:
```diff
@@ -65,6 +66,8 @@ public static void ValidateFilesEqual(string generatedFileName, string reference
+                    codeDiffCommand +
                     "\r\n---------------- To make this the new reference file, execute this ------------------\r\n\r\n    " +
                     tfEditCommand + "    " +
                     copyCommand + "\r\n\r\n-------------------------"
```
same as above regarding tf commands

lindellhugo replied:
OK!

Daniel-Svensson commented on src/Test/Desktop/OpenRiaServices.Common.DomainServices.Test/Baselines/Default/EF/Catalog_EFCore.g.cs:
```diff
@@ -0,0 +1,2697 @@
+        /// Gets or sets the 'ProductModelID' value.
+        /// </summary>
+        [DataMember()]
+        public Nullable<int> ProductModelID
```
Catalog_EFDbCtx.g.cs has "RoundtripOriginal" attribute on this property, but I don't se it explicitly added on the DbContextModel ..

We should investigate why, if it is related to Foreign keys, relations or something else.
There are a few similar differences which I don't understand at a quick glance.

Daniel-Svensson replied:
I am fine with merging this anyway as long as these "todo's/investigations/remaining work" needed are clearly documented   (using issues or README/TODO for the project) to make it clear what is done or not.

Daniel-Svensson replied:
other properties `SizeUnitMeasureCode, WeightUnitMeasureCode, EmployeeID, ShipMethodID, VendorID`,

lindellhugo replied:
Investigated a some of these properties.

They are not flagged with RoundTripOriginal since they are not considered as FK properties and thus will not have the RoundTripAttribute added. This is correct since we for example have decided to exclude both `Vendor` and `UnitMeasure` in the EF Core version of the adventure works model which only contains a subset of the original one. Hence to get the roundtrip attribute we need to extend the model with these entities and reference them.

It still seems to work though, since it was generated correct for for example PurchaseOrderID where both properties was contained in the model.

Added this as a task and added a point in the issue.

Daniel-Svensson reviewed (commented):
some review comments

Daniel-Svensson commented on src/OpenRiaServices.Server.EntityFrameworkCore/Framework/EFCoreDescriptionProvider.cs:
```diff
@@ -0,0 +1,69 @@
+                if (model != null && model.FindEntityType(objectType.FullName) is IEntityType entityType)
+                {
+                    // TODO: ...
+                    // only add an LTE TypeDescriptor if the type is an EF Entity or ComplexType
```
updated

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OpenRIAServices/OpenRiaServices?pullRequest=335) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OpenRIAServices/OpenRiaServices?pullRequest=335) before we can accept your contribution.<br/><hr/>**admin** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenRIAServices/OpenRiaServices?pullRequest=335) it.</sub>

lindellhugo commented:
Add EF Core 3.1 support

- Added EFCore versions of relevant EF based projects
- Added EFCore model for AdventureWorks (partly) and Northwind projects for use in tests
- Added DataAnotations using FluentAPI for the added models

- Added test, notably
  - Added code generation tests (OpenRiaServices.Tools.* test projects)
  - Added code generation tests which EF uses Catalog_EFDbCtx.g.cs and similar tests such as Northwind_EF.g.cs/vb
